src/middlewared/middlewared/plugins/failover_/reboot.py

```diff
@@ -18,72 +28,74 @@ class Config:
+            boot_id = None
 
-        return info
+        self.remote_reboot_reasons[code] = RemoteRebootReason(boot_id, reason)
 
-    @private
-    async def info_impl(self):
-        # initial state
-        current_info = await self.boot_ids()
-        current_info['this_node'].update({'reboot_required': None})
-        current_info['other_node'].update({'reboot_required': None})
-
-        fips_change_info = await self.middleware.call('keyvalue.get', FIPS_KEY, False)
-        if not fips_change_info:
-            for i in current_info:
-                current_info[i]['reboot_required'] = False
-            return current_info
-
-        for key in ('this_node', 'other_node'):
-            current_info[key]['reboot_required'] = all((
-                fips_change_info[key]['id'] == current_info[key]['id'],
-                fips_change_info[key]['reboot_required']
-            ))
-
-        if all((
-            current_info['this_node']['reboot_required'] is False,
-            current_info['other_node']['reboot_required'] is False,
-        )):
-            # no reboot required for either controller so delete
-            await self.middleware.call('keyvalue.delete', FIPS_KEY)
-
-        return current_info
-
-    @accepts(roles=['FAILOVER_READ'])
-    @returns(Dict(
-        'info',
-        Dict('this_node', UUID('id'), Bool('reboot_required')),
-        Dict('other_node', UUID('id', null=True), Bool('reboot_required', null=True)),
-    ))
+        await self.send_event()
+
+    @api_method(FailoverRebootInfoArgs, FailoverRebootInfoResult, roles=['FAILOVER_READ'])
     async def info(self):
-        """Returns the local and remote nodes boot_ids along with their
-        reboot statuses (i.e. does a reboot need to take place)"""
-        return await self.info_impl()
-
-    @accepts(roles=['FAILOVER_READ'])
-    @returns(Bool())
-    async def required(self):
-        """Returns whether this node needs to be rebooted for failover/security
-        system configuration changes to take effect."""
-        # TODO: should we raise Callerror/ValidationError if reboot_required is None?
-        return (await self.info())['this_node']['reboot_required'] is True
-
-    @accepts(roles=['FULL_ADMIN'])
-    @returns()
+        changed = False
+
+        try:
+            other_node = await self.middleware.call('failover.call_remote', 'system.reboot.info', [], {
+                'raise_connect_error': False,
+                'timeout': 2,
+                'connect_timeout': 2,
+            })
+        except Exception:
+            self.logger.warning('Unexpected error querying remote reboot info', exc_info=True)
+            other_node = None
+
+        if other_node is not None:
+            for remote_reboot_reason_code, remote_reboot_reason in list(self.remote_reboot_reasons.items()):
+                if remote_reboot_reason.boot_id is None:
+                    # This reboot reason was added while the remote node was not functional.
```

Comment:
The reason why I used the keyvalue plugin was to avoid double-reboot scenarios like this so I would interpret this new behavior as a regression. Was there a reason why you stopped using keyvalue?

Reply:
You are correct. I added the persistence for `self.remote_reboot_reasons` via `keyvalue` plugin.